internal/db/db.go

```diff
@@ -161,64 +148,431 @@ func InsertVoucher(voucher Voucher) error {
+		}
+		if item.UserInput != "" {
+			enc, err := encodeRvValue(protocol.RVUserInput, item.UserInput)
+			if err != nil {
+				return nil, err
+			}
+			others = append(others, protocol.RvInstruction{Variable: protocol.RVUserInput, Value: enc})
+		}
+		if item.ExtRV != "" {
+			enc, err := encodeRvValue(protocol.RVExtRV, item.ExtRV)
+			if err != nil {
+				return nil, err
+			}
+			others = append(others, protocol.RvInstruction{Variable: protocol.RVExtRV, Value: enc})
+		}
+
+		// grouping is here because of a bug https://github.com/fido-device-onboard/go-fdo/issues/145
+		group := make([]protocol.RvInstruction, 0, len(others)+len(protocols)+len(ports))
+		group = append(group, others...)
+		group = append(group, protocols...)
+		group = append(group, ports...)
+		out = append(out, group)
+	}
+	return out, nil
+}
+
+func parsePortValue(v any) (uint16, error) {
+	switch t := v.(type) {
+	case float64:
+		return uint16(t), nil
+	case string:
+		if t == "" {
+			return 0, fmt.Errorf("empty")
+		}
+		i, err := strconv.Atoi(t)
+		if err != nil {
+			return 0, err
+		}
+		return uint16(i), nil
+	default:
+		return 0, fmt.Errorf("unsupported type %T", v)
+	}
+}
+
+func protocolCodeFromString(s string) (uint8, error) {
+	switch s {
+	case "rest":
+		return uint8(protocol.RVProtRest), nil
+	case "http":
+		return uint8(protocol.RVProtHTTP), nil
+	case "https":
+		return uint8(protocol.RVProtHTTPS), nil
+	case "tcp":
+		return uint8(protocol.RVProtTCP), nil
+	case "tls":
+		return uint8(protocol.RVProtTLS), nil
+	case "coap+tcp":
+		return uint8(protocol.RVProtCoapTCP), nil
+	case "coap":
+		return uint8(protocol.RVProtCoapUDP), nil
+	default:
+		return 0, fmt.Errorf("unsupported protocol %q", s)
+	}
+}
+
+func parseMediumValue(v any) (uint8, error) {
+	switch t := v.(type) {
+	case float64:
+		return uint8(t), nil
+	case string:
+		switch t {
+		case "eth_all":
+			return protocol.RVMedEthAll, nil
+		case "wifi_all":
+			return protocol.RVMedWifiAll, nil
+		default:
+			return 0, fmt.Errorf("unsupported medium %q", t)
+		}
+	default:
+		return 0, fmt.Errorf("unsupported medium type %T", v)
+	}
+}
+
+// ParseHumanToTO2AddrsJSON parses a JSON like
+// [{"dns":"fdo.example.com","port":"8082","protocol":"http","ip":"127.0.0.1"}]
+// into []protocol.RvTO2Addr.
+func parseHumanToTO2AddrsJSON(rawJSON []byte) ([]protocol.RvTO2Addr, error) {
+	// Strongly-typed JSON for validation
+	type to2Human struct {
+		DNS      string `json:"dns"`
+		IP       string `json:"ip"`
+		Port     string `json:"port"`
+		Protocol string `json:"protocol"`
+	}
+	var items []to2Human
+	if err := json.Unmarshal(rawJSON, &items); err != nil {
+		return nil, fmt.Errorf("invalid TO2 addrs JSON: %w", err)
+	}
+
+	out := make([]protocol.RvTO2Addr, 0, len(items))
+	for i, item := range items {
+		var (
+			ipPtr  *net.IP
+			dnsPtr *string
+			port   uint16
+			proto  protocol.TransportProtocol
+		)
+
+		if item.IP != "" {
+			ip := net.ParseIP(item.IP)
```

Comment:
check for error and return error , if invalid IP then ParseIP will return nil and we might end up saving nil IP address in database.